runtime/onert/backend/train/TensorPlanner.cc

```diff
@@ -35,9 +35,184 @@ TensorPlanner::TensorPlanner(const ir::train::TrainableGraph &tgraph,
+  std::vector<ir::train::TrainingOperandIndex> operands_last_until_end;
+  for (const auto &pair : uses_map)
+  {
+    const auto &operand_index = pair.first;
+    const auto use_count = pair.second;
+    if (use_count == 0)
+      operands_last_until_end.push_back(operand_index);
+  }
+
+  // Plan used or defined tensors in forwarding nodes
+  // At each operation,
+  // 1. Scan DEF of outputs. If the DEF, allocate it
+  // 2. Scan DEF of inputs. If variable tensor, throw an exception (not supported yet)
+  // 3. Scan USE of inputs/outputs. Decrease the USE and deallocate if the USE is 0
+  const auto order = _tgraph.topolSortOperations();
+  for (const auto &op_index : order)
+  {
+    const auto &op = _tgraph.operations().at(op_index);
+    auto op_inputs = op.getInputs() | ir::Remove::DUPLICATED | ir::Remove::UNDEFINED;
+    auto op_outputs = op.getOutputs() | ir::Remove::DUPLICATED | ir::Remove::UNDEFINED;
+
+    // Define outputs
+    for (const auto &output : op_outputs)
+    {
+      if (_external_operands.contains(output))
+        continue;
+      if (!tensor_builder->isRegistered(output))
+        continue;
+
+      const auto output_index = ir::train::TrainingOperandIndex{output, true};
+      assert(defs_map.find(output_index) != defs_map.end());
+      defs_map[output_index] = 0;
+      tensor_builder->notifyFirstUse(output_index.index());
+    }
+
+    // Scan variable tensors
+    // This tensor has features like constant. But OperandInfo and LowerInfo treat them as
+    // non-constant because of less memory usage by memory planning in here
+    // However, train backend does not support variable tensors yet
+    for (const auto &input : op_inputs)
+    {
+      if (_external_operands.contains(input))
+        continue;
+      if (!tensor_builder->isRegistered(input))
+        continue;
+
+      const auto input_index = ir::train::TrainingOperandIndex{input, true};
+      const auto &operand = training_usedefs.at(input_index).operand();
+      if (operand.isConstant())
+        continue;
+
+      assert(training_usedefs.find(input_index) != training_usedefs.end());
+      if (operand.info().isVariable())
+        throw std::runtime_error("The train backend does not support variable tensors");
+    }
+
+    for (const auto &input : op_inputs)
+    {
+      if (_external_operands.contains(input))
+        continue;
+      if (!tensor_builder->isRegistered(input))
+        continue;
+
+      const auto input_index = ir::train::TrainingOperandIndex{input, true};
+      const auto &operand = training_usedefs.at(input_index).operand();
+      if (operand.isConstant())
+        continue;
+
+      assert(uses_map.find(input_index) != uses_map.end());
+      assert(uses_map[input_index] > 0);
+      uses_map[input_index]--;
+      if (uses_map[input_index] == 0)
+      {
+        // plan for deallocation of static tensor node
+        tensor_builder->notifyLastUse(input_index.index());
+      }
+    }
+  }
```

Comment:
This code is similar to https://github.com/Samsung/ONE/blob/19c259d2b462bf260addc4a24f0b3590cf33c1cf/runtime/onert/core/include/backend/basic/BackendContextHelpers.h#L103-L163